### ISSUE TITLE
Missing scripts/ and train files

### ISSUE THREAD

**Issue by akhilgakhar** [RGenDiff/open-litevae#1]
Thankyou for creating the code implementation of the paper. 
Its an interesting read!

Can you also share the scripts folder and code for training as well from scratch :)
Thankyou

**Comment by vasimr** [comment]
Thank you for your interest.

The dataset, evaluation, and training code have yet to be detached (refactored) from a larger codebase which cannot be shared. Out of the three, the training code is the most challenging since it relies on an older version of lightning and should probably be updated. 
The plan is to work on this process sometime next month since the LiteVAE subset is also being used for an upcoming conference submission.

If you plan to use the code for a project before then, you can adapt the [TamingTransformers](https://github.com/CompVis/taming-transformers) codebase, which is what this implementation is based on. I can also post the Gaussian Kernel and Wavelet loss functions in a code snippet if needed.

**Comment by akhilgakhar** [comment]
Thankyou for your generous reply.

I understand. I will look into the mentioned codebase.
In the meanwhile, if possible, please provide Gaussian Kernel and Wavelet loss functions in some code snippet :)
If there is anything else that can be shared, kindly do the needful

**Comment by vasimr** [comment]
Bellow are the new loss kernels.

```
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch_dwt.functional import dwt2

##########################################
##        New Loss methods for LiteVAE
##########################################

class GaussianFilter(nn.Module):
    def __init__(self, kernel_size=5, sigma=1.0):
        super(GaussianFilter, self).__init__()
        self.kernel_size = kernel_size
        self.sigma = sigma
        # Create the Gaussian kernel
        self.kernel = self.create_gaussian_kernel(kernel_size, sigma)
        # Register as a buffer, ensuring it's not a parameter and won't be updated
        self.register_buffer('gaussian_kernel', self.kernel)

    def create_gaussian_kernel(self, size, sigma):
        """Create a Gaussian kernel for convolution."""
        x = torch.arange(-size // 2 + 1, size // 2 + 1, dtype=torch.float32)
        g = torch.exp(-x.pow(2) / (2 * sigma ** 2))
        g = g[:, None] * g[None, :]
        g /= g.sum()
        return g

    def forward(self, x):
        """Apply the Gaussian filter to the input tensor x."""
        channels = x.size(1)
        # Ensure the kernel is correctly shaped [out_channels, in_channels, H, W]
        kernel = self.gaussian_kernel[None, None, :, :].repeat(channels, 1, 1, 1)
        padding = self.kernel_size // 2
        return F.conv2d(x, kernel, padding=padding, groups=channels)

def charbonnier_loss(prediction, target, epsilon=1e-3):
    # Calculate the difference between the predicted and target values
    diff = prediction - target

    # Compute the Charbonnier loss
    loss = torch.sqrt((diff ** 2) + (epsilon ** 2))

    return loss

def compute_wavelet_loss(x, x_hat):
    # Decompose both images
    _, x_LH, x_HL, x_HH = dwt2(x, "haar").unbind(dim=1)
    _, x_hat_LH, x_hat_HL, x_hat_HH = dwt2(x_hat, "haar").unbind(dim=1)
    
    # Compute Charbonnier loss for each high-frequency component
    loss_LH = charbonnier_loss(x_LH, x_hat_LH)
    loss_HL = charbonnier_loss(x_HL, x_hat_HL)
    loss_HH = charbonnier_loss(x_HH, x_hat_HH)
    
    # Average the losses
    total_loss = (loss_LH + loss_HL + loss_HH) / 3
    return total_loss

```

And this is how they are applied with the VAE loss:

```
if self.gaussian_weight > 0:
    gk_loss = self.compute_gaussian_loss(inputs.contiguous(), reconstructions.contiguous())

if self.wavelet_weight > 0:
    with autocast(enabled=False): # disable autocast since wavelets use conv3D - also need to convert to float32
        wv_loss = compute_wavelet_loss(inputs.float().contiguous(), reconstructions.float().contiguous())
        wv_loss = wv_loss.mean(dim=(-1,-2), keepdim=True) # collapse with H,W using mean since h=H//2, w=W//2
        # when adding back to the full B,C,H,W tensor, this should have a similar effect to preserving the magnitude
        # i.e. sum_N x + (w0 + w1 + w2 + w3 + ...)/N ~ sum_N x + sum_N w

# sum the nll loss terms
nll_loss = rec_loss + self.wavelet_weight*wv_loss + self.gaussian_weight*gk_loss
```
Finally, below are the discriminator adapters, which handle the case where multiple values are returned (i.e. for GigaGAN and UNet-GAN). 

```
##########################################
##         GAN LOSS Wrappers
##########################################

# wrapper functions for flattening the loss types
def g_loss_wrappper(logits_fake, loss_fn):
    if isinstance(logits_fake, torch.Tensor):
        loss = loss_fn(logits_fake)
    else:
        loss = 0
        for fi in logits_fake:
            loss = loss + loss_fn(fi.view(-1))
        # normalize
        loss = loss/len(logits_fake)

    return loss

def d_loss_wrappper(logits_real, logits_fake, loss_fn):
    if isinstance(logits_real, torch.Tensor):
        loss = loss_fn(logits_real, logits_fake)
    else:
        loss = 0
        for ri, fi in zip(logits_real, logits_fake):
            loss = loss + loss_fn(ri.view(-1), fi.view(-1))
        # normalize
        loss = loss/len(logits_real)

    return loss
```

The easiest way to use them is via a function pointer in the loss initializer. 
```
from .losses import get_gan_loss

# get the GAN loss functions, SD and LiteVAE use disc_loss='hinge'
g_func, d_func = get_gan_loss(disc_loss) 

# wrap the loss functions to handle tuple responses
self.gen_loss = partial(g_loss_wrappper, loss_fn=g_func)
self.disc_loss = partial(d_loss_wrappper, loss_fn=d_func)
```

Unfortunately it looks like I may not be able to work on cleaning and uploading the full code until June'25 (too many conference deadlines). I hope this helps in the mean time!

**Comment by akhilgakhar** [comment]
Thank a lot <3
It helps.

**Comment by BOR54** [comment]
Thanks for the snippets. They're really helpful. Also, I just checked your proposed timeline and its seems you might have finished revising the code. Are you able to share the updated training script? If not,  are you able to share the training script as it is with the older lightning?
